// api/lib/apis/account.js
/**
 * Created by 沐沐 on 2015-10-14.
 * 账号接口
 */
var request = require('./request');


/**
 * 重置手机号码
 * @param tokenOrUnionid
 * @param {string} captcha 第二次的验证码
 * @param {string} captchaOld 第一次的验证码
 * @param {string} mobile 手机号码
 * @returns {*}
 */
module.exports.updateMobile = function (tokenOrUnionid, captcha, captchaOld, mobile) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.local.siebel.member.update',
            captcha: captcha,
            captchaOld: captchaOld,
            mobile: mobile
        }, tokens);
    });
};


/**
 * 设置支付密码
 * @param tokenOrUnionid
 * @param cipherKey 加密串
 * @param payPwd 支付密码
 * @returns {*}
 */
module.exports.setPayPass = function (tokenOrUnionid, cipherKey, payPwd) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.bbgpay.member.set.password.wechat',
            //cipherKey: cipherKey,
            payPwd: payPwd
        }, tokens);
    });
};

/**
 * 修改支付密码
 * @param tokenOrUnionid
 * @param cipherKey 新支付密码加密串
 * @param oriCipherKey 原支付密码加密串
 * @param payPass 新支付密码
 * @param oriPayPass 原支付密码
 * @param code 短信验证码
 * @param phone 手机号码
 * @returns {*}
 */
module.exports.updatePayPass = function (tokenOrUnionid, cipherKey, oriCipherKey, payPass, oriPayPass, code, phone) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.paypass.update.wechat',
            //cipherKey: cipherKey,
            //oriCipherKey: oriCipherKey,
            payPass: payPass,
            oriPayPass: oriPayPass,
            code: code,
            phone: phone
        }, tokens)
    });
};

/**
 * 重置手机号码申述
 * @param {string} origPhone 原手机号码
 * @param {string} newPhone 修改后手机号码
 * @param {string} identificationNum 身份证号码
 * @param {string} fileName1
 * @param {string} fileName2
 * @param image1 身份证正面图片文件
 * @param image2 身份证背面图片文件
 * @param {string} smscode 短信验证码
 */
module.exports.userAppealPhone = function (origPhone, newPhone, identificationNum, image1, image2, fileName1, fileName2, smscode) {
    return request.post({
        method: 'bubugao.mobile.userappeal.phone',
        origPhone: origPhone,
        newPhone: newPhone,
        identificationNum: identificationNum,
        fileName1: fileName1,
        fileName2: fileName2,
        smscode: smscode
    }, {
        image1: image1,
        image2: image2
    }, true)
};

/**
 * 修改支付密码申述
 * @param {string | object} tokenOrUnionid
 * @param {string} payPwd 支付密码
 * @param {string} cipherKey 支付密码加密串
 * @param {string} identificationNum 身份证号码
 * @param image1 身份证正面图片文件
 * @param image2 身份证背面图片文件
 * @param image3 手持身份图片文件
 * @param {string} fileName1 身份证正面文件名称
 * @param {string} fileName2 身份证背面文件名称
 * @param {string} fileName3 手持身份证文件名称
 * @param {string} smscode 短信验证码
 */
module.exports.userAppealPayPass = function (tokenOrUnionid, payPwd, cipherKey, identificationNum, image1, image2, image3, fileName1, fileName2, fileName3, smscode) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.postAuth({
            method: 'bubugao.mobile.userappeal.paypass',
            payPwd: payPwd,
            cipherKey: cipherKey,
            identificationNum: identificationNum,
            fileName1: fileName1,
            fileName2: fileName2,
            fileName3: fileName3,
            smscode: smscode
        }, {
            image1: image1,
            image2: image2,
            image3: image3
        }, tokens, true);
    });
};

/**
 * 支付密码申述状态
 * @param tokenOrUnionid
 * @returns status：0是未申诉， 1正在申诉
 */
module.exports.userAppealPayPassStatus = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.userappeal.paypass.status'
        }, tokens);
    });
};

/**
 * App2.6发送短信验证码接口
 * @param {string} mobile 手机号码
 * @param {string} type = [login|warning|pay|mobile|update_mobile|update_pay_password|update_mobile_complaint] 短信类型
 * @returns {promise}
 */
module.exports.sms = function (mobile, type) {
    return request.get({
        method: 'bubugao.mobile.register.login.mobile.sms.send',
        mobile: mobile,
        type: type
    });
};

/**
 * 验证手机短信验证码
 * @param {string} captcha
 * @param {string} mobile
 * @param {string} smsType = [login|warning|pay|mobile|update_mobile|update_pay_password|update_paypass_complaint]
 * @returns {promise}
 */
module.exports.smsVerify = function (captcha, mobile, smsType) {
    return request.get({
        method: 'bubugao.mobile.verify.mobile.sms',
        captcha: captcha,
        mobile: mobile,
        smsType: smsType
    });
};

/**
 * 收货地址查询
 * @param tokenOrUnionid
 * @returns {*}
 */
module.exports.address = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        // return request.getAuth({
        //     method: 'bubugao.mobile.mall.member.address.search'
        // }, tokens);

        return {
            code:'0000',
            data:[],
            message:'接口已屏蔽'
        }
    });
};

/**
 * 自助积分扫码接口(梅溪湖项目新增接口)
 * @param tokenOrUnionid
 * @param cardNo 积分卡号码
 */
module.exports.pointIncrease = function (tokenOrUnionid, cardNo) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.integralstore.increase.point',
            cardNo: cardNo
        }, tokens).fail(function(err){
            if(err.code === '9999'){
                err.code = '9998';
            }
            throw err;
        });
    });
};
