// api/lib/apis/assistance.js
/**
 * Created by 沐沐 on 2015-09-15.
 * 助力
 */
var request = require('./request');


module.exports.join = function (openId, nickname, headImage, name, school, mobile) {
    return request.get({
        method: 'bubugao.mobile.wechat.assistance.join',
        openId: openId,
        nickname: nickname,
        headImage: headImage,
        name: name,
        school: school,
        mobile: mobile
    });
};

module.exports.school = function () {
    return request.get({
        method: 'bubugao.mobile.wechat.assistance.school'
    });
};

module.exports.assistance = function (identityCode, openId) {
    return request.get({
        method: 'bubugao.mobile.wechat.assistance.assistance',
        assistanceId: openId,
        byAssistanceCode: identityCode
    });
};

module.exports.canAssistance = function (identityCode, openId) {
    return request.get({
        method: 'bubugao.mobile.wechat.assistance.can.assistance',
        assistanceId: openId,
        byAssistanceCode: identityCode
    });
};

module.exports.getIdentityCode = function (openId) {
    return request.get({
        method: 'bubugao.mobile.wechat.assistance.getcode.by.openid',
        openId: openId
    }).then(function (identityCode) {
        return identityCode.data.identityCode;
    });
};

module.exports.find = function (identityCode) {
    return request.get({
        method: 'bubugao.mobile.wechat.assistance.getjoin.by.code',
        identityCode: identityCode
    });
};

module.exports.top = function (identityCode) {
    return request.get({
        method: 'bubugao.mobile.wechat.assistance.top',
        identityCode: identityCode
    });
};
