// api/lib/apis/coupons.js
/**
 * Created by 沐沐 on 2015-07-30.
 * 优惠券接口
 * Last update time by rico on 2016-11-10
 */

var request = require('./request');
var q = require('q');

/**
 * 商城优惠券
 * @param tokenOrUnionid
 * @param page
 * @param pageSize
 * @param status
 * @returns {promise}
 */
module.exports.couponMall = function (tokenOrUnionid, status, page, pageSize) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        // return request.getAuth({
        //     method: 'bubugao.mobile.mall.ump.coupon.get',
        //     pageNo: page,
        //     pageSize: pageSize,
        //     status: status//可用
        // }, tokens).then(function (data) {
        //     return {
        //         count: data.data.data.length,
        //         total: data.data.totalPages,
        //         index: data.data.pageNum,
        //         data: data.data.data.map(function (e) {
        //             var effectStartTime = new Date(e.effectStartTime), effectEndTime = new Date(e.effectEndTime);
        //             return {
        //                 name: e.cpnsName,//优惠券名称
        //                 info: e.info,//备注信息
        //                 activeTime: e.activeTime,//不知道啥含义
        //                 startTime: effectStartTime.getFullYear() + '-' + (effectStartTime.getMonth() + 1) + '-' + effectStartTime.getDate(),//开始时间
        //                 endTime: effectEndTime.getFullYear() + '-' + (effectEndTime.getMonth() + 1) + '-' + effectEndTime.getDate(),//结束时间
        //                 type: e.type,//1,现金券.2,折扣券
        //                 amount: e.amount / 100,//抵扣现金
        //                 discount: (e.discount / 10).toFixed(1),//折扣
        //                 shopName: e.shopName,//商店名称
        //                 shopId: e.shopId,//商店id
        //                 code: e.code, //券码
        //                 memberId: e.memberId,//电商会员编号
        //                 status: e.status//状态
        //             };
        //         })
        //     };
        // }, function (err) {
        //     if (err.code === '1010') {
        //         throw err;
        //     } else {
        //         return {
        //             count: 0,
        //             total: 0,
        //             index: 0,
        //             data: []
        //         }
        //     }
        // });
        return {
            count: 0,
            total: 0,
            index: 0,
            data: []
        }
    });
};

/**
 * 联盟商户优惠券
 * @param tokenOrUnionid
 * @param page
 * @param pageSize
 * @param status 1：未使用 ， 2：已使用 ，3：已失效，4：未生效"
 * @returns {promise}
 */
module.exports.couponMember = function (tokenOrUnionid, status, page, pageSize) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.coupcons.member.get',
            page: page,
            pageSize: pageSize,
            status: status//可用
        }, tokens).then(function (data) {
            return {
                count: data.data.length,
                total: data.total,
                index: data.page,
                data: data.data.map(function (e) {
                    return {
                        id: e.id,//主键
                        couponid: e.couponid,//优惠券主键
                        name: e.name,//优惠券名称
                        status: e.status,//优惠券状态 1：未使用 ， 2：已使用 ，3：已失效，4：未生效
                        startTime: e.begintime.split(' ')[0],//开始日期
                        endTime: e.endtime.split(' ')[0],//结束日期
                        image: e.image,//优惠券图片
                        code: e.code,//优惠券使用编码
                        auditstatus: e.auditstatus,//优惠券审批状态
                        stores: e.stores//门店
                    };
                })
            };
        });
    });
};

/**
 * 联盟优惠券详情
 * @param couponId
 * @returns {promise}
 */
module.exports.couponMemberDetail = function (couponId) {
    return request.get({
        method: 'bubugao.mobile.coupcons.detail',
        couponId: couponId
    });
};

/**
 * 门店优惠券
 * @param tokenOrUnionid
 * @param page
 * @param pageSize
 * @param status (已取消 ;可用;已应用;过期的)
 * @returns {promise}
 */
module.exports.couponSiebel = function (tokenOrUnionid, status, page, pageSize) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.local.siebel.vouchers.get',
            page: page,
            pageSize: pageSize,
            status: status
        }, tokens).then(function (data) {
            return {
                count: data.data.length,
                total: data.total,
                index: data.page,
                data: data.data.map(function (e) {
                    return {
                        voucherNo: e.voucherNo,//
                        voucherName: e.voucherName || '',//优惠券类型
                        vendorAmt: e.vendorAmt,//优惠券金额
                        cashFlag: e.cashFlag,//是否现金券
                        startTime: e.startDt,//优惠券开始时间
                        endTime: e.endDt,//优惠券结束时间
                        usedDt: e.usedDt || '',//优惠券使用时间
                        vendorDesc: e.vendorDesc || '',//优惠券说明
                        status: e.status,//状态
                        dismissVchPartner: e.dismissVchPartner || '',//发优惠券交易的业态
                        useVchPartner: e.useVchPartner || '',//使用优惠券交易的业态
                        useVchCarrier: e.useVchCarrier || '',//使用优惠券交易的门店
                        fourCheckNo: e.fourCheckNo//优惠券编号
                    }
                })
            };
        });
    });
};


/**
 * 第三方优惠券
 * @param tokenOrUnionid
 * @param page
 * @param pageSize
 * @param status  status =0 已过期，status =1 可使用
 * @returns {*}
 */
module.exports.couponThird = function (tokenOrUnionid, page, pageSize, status) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.coupcons.thridkeys.list',
            page: page,
            pageSize: pageSize,
            status: status
        }, tokens).then(function (data) {
            return {
                count: data.data.length,
                total: data.total,
                index: data.page,
                data: data.data.map(function (e) {
                    return {
                        name: e.name,
                        image: e.image,
                        startTime: e.begintime,
                        endTime: e.endtime,
                        voucherNo: e.voucherNo
                    }
                })
            };
        });
    });
};


/**
 * 第三方优惠券详情
 * @param tokenOrUnionid
 * @param giftId
 * @param code
 * @returns {*}
 */
module.exports.couponThridDetail = function (tokenOrUnionid, giftId, code) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.coupcons.thridkeys.detail',
            code: code,
            giftId: giftId
        }, tokens).then(function (data) {
            return data;
        });
    });
};

/**
 * 优惠券统计
 * @param tokenOrUnionid
 * @returns {*}
 */
module.exports.countAll = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        //console.log(tokens);
        return q.all([
            module.exports.couponMall(tokens, 1, 1, 1),
            // module.exports.couponMember(tokens, 1, 1, 1),
            module.exports.couponSiebel(tokens, '可用', 1, 1),
            module.exports.couponThird(tokens, 1, 1, 1)
        ]).spread(function (mall, siebal, third) {
            return {
                mall: mall.total,
                // member: member.total,
                siebal: siebal.total,
                third: third.total
            };
        });
    });
};


/**
 * siebel 优惠券列表 非登录
 * @param storeid 不传则表示大会员默认频道id
 * @param type 优惠券类型 0超市 4百货 5母婴 6电器 7停车
 * @param sort  1:领取次数排序 2：价格排序 3：时间排序 为空不传都默认按时间排序
 * @param page
 * @param pageSize
 * @returns {*}
 */
module.exports.couponList = function (storeid,type,sort,page, pageSize) {
    return request.get({
        method: 'bubugao.mobile.coupons.list',
        storeid:storeid,
        type:type,
        sort:sort,
        page: page,
        pageSize: pageSize
    });
};

/**
 * siebel 优惠券用户领取列表 登录
 * @param tokenOrUnionid
 * @param storeid 不传则表示大会员默认频道id "H0101"
 * @param type 优惠券类型 0超市 4百货 5母婴 6电器 7停车
 * @param sort  1:领取次数排序 2：价格排序 3：时间排序 为空不传都默认按时间排序
 * @param page
 * @param pageSize
 * @returns {*}
 */
module.exports.couponAuthList = function (tokenOrUnionid,storeid,type,sort,page, pageSize,tokenid) {
    return request.getTokens(tokenOrUnionid,tokenid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.coupons.list.login',
            storeid:storeid,
            type:type,
            sort:sort,
            page: page,
            pageSize: pageSize
        }, tokens).then(function (data) {
            return data;
        });
    });
};

/**
 * 优惠卷领取
 * @param tokenOrUnionid
 * @param storeid   门店id  不传则表示大会员默认频道id
 * @param couponId 必须  优惠券id
 * @param carNo 车牌号
 * @returns {*}
 */
module.exports.couponReceiveNew = function (tokenOrUnionid, storeid,couponId, carNo, tokenid) {
    return request.getTokens(tokenOrUnionid,tokenid).then(function (tokens) {
        var _param={
            method: 'bubugao.mobile.coupcons.receive.new',
            couponId: couponId,
            storeid:storeid
        };
        if(carNo && carNo.length>0){
            _param.car_no=carNo;
        }
        return request.getAuth(_param, tokens).then(function (data) {
            return data;
        });
    });
};
