// api/lib/apis/ferriswheel.js
/**
 * Created by 沐沐 on 2015-07-31.
 */

var request = require('./request');
var weixin = require('../weixin');
var area = require('./area');
var user = require('./user');
var q = require('q');

var activeCode = 'chb';

module.exports.shareMessage = function (id) {
    return request.get({
        method: 'bubugao.mobile.app.party.sharemessage.get',
        type: id
    }).then(function (data) {
        return data.data;
    });
};

/**
 * 摩天轮奖品
 * @param tokenOrUnionid
 * @param area
 * @returns {promise}
 */
module.exports.goods = function (tokenOrUnionid, area) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.lottery.query4goods',
            code: activeCode,
            cityId: area
        }, tokens);
    });
};

/**
 * 剩余机会
 * @param tokenOrUnionid
 * @returns {promise}
 */
module.exports.time = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.lottery.times.get',
            code: activeCode
        }, tokens);
    });
};

/**
 * 中奖纪录
 * @param activityid
 * @returns {*}
 */
module.exports.record = function (activityid) {
    return request.get({
        method: 'bubugao.mobile.lottery.query4records',
        activityId: activityid,
        count: 20
    }).then(function (data) {
        return data.data.map(function (e) {
            return e.lottTimeDesc + '，' + e.pseudoName + '抽中' + e.goodsName;
        });
    });
};

/**
 * 摩天轮规则
 * @param activityid
 * @param area
 * @returns {promise}
 */
module.exports.rules = function (activityid, area) {
    return request.get({
        method: 'bubugao.mobile.lottery.rule',
        activityId: activityid,
        cityId: area
    }).then(function (data) {
        return data.data.detail
    });
};

/**
 * 初始化参数
 * @param tokenOrUnionid
 * @param url
 * @returns {*}
 */
module.exports.init = function (tokenOrUnionid, url) {
    return weixin.sign(url).then(function (sign) {
        return {
            sign: sign
        }
    })
};


var getInfo = function (tokenOrUnionid, areaCode) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return module.exports.goods(tokens, areaCode).then(function (goods) {
            return q.all([
                module.exports.record(goods.data.activityid),
                module.exports.rules(goods.data.activityid, areaCode),
                module.exports.shareMessage(3),//摩天轮编号
                user.mobile(tokens),
                area.getRegionByCode(areaCode, 1)
            ]).spread(function (record, rules, shareMessage, mobile, areaObj) {
                return {
                    goods: goods.data.lotteryGoods,
                    time: goods.data.times,
                    records: record,
                    rules: rules,
                    activityid: goods.data.activityid,
                    area: areaObj.data,
                    shareMessage: shareMessage,
                    mobile: mobile.data.mobile
                };
            })
        });
    });
};

/**
 * 初始化信息
 * @param tokenOrUnionid
 * @param lat
 * @param lng
 * @param code
 * @returns {*}
 */
module.exports.info = function (tokenOrUnionid, lat, lng, code) {
    var defaultArea = code || '430100000000';
    if (lat && lng) {
        return q.all([
            request.getTokens(tokenOrUnionid),
            area.area(lat, lng)
        ]).spread(function (tokens, areaCode) {
            areaCode = (areaCode && areaCode.id) || defaultArea;
            return getInfo(tokens, areaCode);
        });
    } else {
        return request.getTokens(tokenOrUnionid).then(function (tokens) {
            return getInfo(tokens, defaultArea);
        });
    }
};

/**
 * 分享抽奖增加次数
 * @param tokenOrUnionid
 * @param activityid
 * @returns {*}
 */
module.exports.share = function (tokenOrUnionid, activityid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        var time = function () {
            return module.exports.time(tokens);
        };
        return request.getAuth({
            method: 'bubugao.mobile.lottery.create',
            activityId: activityid,
            code: activeCode,
            channel: 'SHARE'
        }, tokens).then(time, time);
    });
};

/**
 * 抽奖
 * @param tokenOrUnionid
 * @param activityid
 * @returns {*}
 */
module.exports.draw = function (tokenOrUnionid, activityid, area) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.lottery.draw.alt',
            activityId: activityid,
            cityId: area
        }, tokens);
    });
};
