// api/lib/apis/gift.js
/**
 * Created by 沐沐 on 2015-10-14.
 * 礼品接口
 */


var request = require('./request');
var user = require('./user');
var q = require('q');


/**
 * * 商品查询
 * @param areaId 默认0为全部
 * @param categoryid 礼品分类ID 默认0为全部
 * @param storesid
 * @param merchantid
 * @param page
 * @param pageSize
 * @param order 排序方式 price 价格, sell 销量
 * @param ordertype  排序方式 asc 升序, desc 降序
 * @param activityType
 * @param isfilterNoBegin
 * @param isfilterNoStore
 * @param beginTime
 * @param giftName
 * @returns {promise}
 */
module.exports.giftSearch = function (areaId, categoryid, storesid, merchantid, page, pageSize, order, orderType, activityType, isfilterNoBegin, isfilterNoStore, beginTime, giftName) {
    return request.get({
        method: 'bubugao.mobile.gift.search',
        areaId: areaId,
        categoryid: categoryid,
        storesid: storesid,
        merchantid: merchantid,
        page: page,
        pageSize: pageSize,
        order: order,
        ordertype: orderType,
        activityType: activityType,
        isfilterNoBegin: isfilterNoBegin,
        isfilterNoStore: isfilterNoStore,
        beginTime: beginTime,
        giftName: giftName
    });
};

/**
 * 根据id串查询商品列表
 * @param giftIds
 * @returns {promise}
 */
module.exports.giftSearchByIds = function (giftIds) {
    return request.get({
        method: 'bubugao.mobile.gift.ids.search',
        giftIds: giftIds
    });
};

/**
 * 按组查询商品列表
 * @param areaId
 * @param page
 * @param pageSize
 * @param spec 图片规格
 * @returns {promise}
 */
module.exports.giftSearchByGroup = function (areaId, page, pageSize, spec) {
    return request.get({
        method: 'bubugao.mobile.gift.group',
        areaId: areaId,
        page: page,
        pageSize: pageSize,
        spec: spec
    });
};

/**
 * 商品详情
 * @param giftid
 * @returns {*|promise}
 */
module.exports.giftDetail = function (giftid) {
    return request.get({
        method: 'bubugao.mobile.gift.detail',
        giftid: giftid
    });
};

/**
 * 商品分类
 * @returns {*|promise}
 */
module.exports.giftCategory = function () {
    return request.get({
        method: 'bubugao.mobile.gift.category.list.get'
    });
};

/**
 * 门店
 * @param isfilterNoStore
 * @param isfilterNoBegin
 * @param areaId
 * @param page
 * @param pageSize
 * @param giftId
 * @returns {*|promise}
 */
module.exports.store = function (isfilterNoStore, isfilterNoBegin, areaId, page, pageSize, giftId) {
    return request.get({
        method: 'bubugao.mobile.store.area',
        _isfilterNoStore: isfilterNoStore,
        areaId: areaId,
        giftId: giftId,
        isfilterNoBegin: isfilterNoBegin,
        page: page,
        pageSize: pageSize
    });
};

/**
 * 查询商品库存
 * @param storeid
 * @param giftid
 * @returns {*|promise}
 */
module.exports.inventory = function (storeid, giftid) {
    return request.get({
        method: 'bubugao.mobile.gift.inventory',
        storeid: storeid,
        giftid: giftid
    });
};

module.exports.invAndMember = function (unionidOrTokens, storeid, giftid) {
    return request.getTokens(unionidOrTokens).then(function (tokens) {
        return q.all([
            module.exports.inventory(storeid, giftid),
            user.meberInfo(tokens)
        ]);
    }).spread(function (inventory,member) {
        inventory.data.member = member;
        return inventory;
    });
};

/**
 * 兑换礼品
 * @param unionidOrTokens
 * @param giftid 商品id
 * @param storeid 门店id
 * @param receiver 收货人
 * @param tel 收货人电话
 * @param address 收货人地址
 * @param distributtype 配送方式  1:到店领取  2:积分/高币付邮 3:到货付邮
 * @param num 订单数量
 * @param pay 支付方式（0 积分，1:高币;2:高币红包; 默认积分,多个以,分隔）
 * @param password 支付密码(支付方式为积分时不需要)
 * @param cipherKey 支付密码秘钥(支付方式为积分时不需要)
 * @returns {*}
 */
module.exports.createOrder = function (unionidOrTokens, giftid, storeid, receiver, tel, address, distributtype, num, pay, password, cipherKey) {
    return request.getTokens(unionidOrTokens).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.gift.order.create.alt',
            giftid: giftid,
            storeid: storeid,
            receiver: receiver,
            tel: tel,
            address: address,
            distributtype: distributtype,
            num: num,
            pay: pay,
            password: password,
            cipherKey: cipherKey
        }, tokens)
    })
};

/**
 * 礼品微信预订单下单接口
 * @param unionidOrTokens
 * @param openid
 * @param distributtype
 * @param tel
 * @param address
 * @param receiver
 * @param giftid
 * @param storeid
 * @param num
 * @returns {*}
 */
module.exports.prepay = function (unionidOrTokens, openid, distributtype, tel, address, receiver, giftid, storeid, num) {
    return request.getTokens(unionidOrTokens).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.gift.order.weichatpay.prepay',
            openid: openid,
            tel: tel,
            address: address,
            distributtype: distributtype,
            receiver: receiver,
            giftid: giftid,
            storeid: storeid,
            num: num
        }, tokens);
    });
};
