// api/lib/apis/hongbao.js
/**
 * Created by 沐沐 on 2015-08-11.
 */
var request = require('./request');
var weixin = require('../weixin');
var q = require('q');

var gameCode = 'yth';

module.exports.shareMessage = function () {
    return request.get({
        method: 'bubugao.mobile.app.party.sharemessage.get',
        type: 7
    });
};

module.exports.draw = function (user) {
    return request.get({
        method: 'bubugao.mobile.lottery.wechat.draw',
        code: gameCode,
        unionid: user.unionid,
        openid: user.openid
    });
};

module.exports.rule = function () {
    return request.get({
        method: 'bubugao.mobile.licenses.get',
        code: 'WECHAT'
    });
};

module.exports.submit = function (unionid, orderid, mobile) {
    return request.get({
        method: 'bubugao.mobile.lottery.wechat.phone.submit',
        code: gameCode,
        orderid: orderid,
        mobile: mobile,
        unionid: unionid
    });
};

module.exports.times = function (user) {
    return request.get({
        method: 'bubugao.mobile.lottery.times.get',
        code: gameCode,
        unionid: user.unionid
    });
};

module.exports.last = function (user) {
    return request.get({
        method: 'bubugao.mobile.lottery.wechat.last.result',
        code: gameCode,
        unionid: user.unionid
    });
};

module.exports.init = function (code, url) {
    return weixin.open_id(code).then(function (user) {
        return q.all([
            module.exports.last(user),
            module.exports.rule(),
            module.exports.shareMessage(),
            weixin.sign(url)
        ]).spread(function (last, rule, shareMessage, sign) {
            return {
                last: {
                    time: last.data.times,
                    mobile: last.data.mobile,
                    lastmobile: last.data.lastmobile,
                    orderid: last.data.orderid,
                    goodsName: last.data.goods.goodsName,
                    goodsMoney: last.data.goods.gaobi
                },
                share: sign,
                shareMessage: {
                    title: shareMessage.data.message,
                    link: shareMessage.data.link,
                    imgUrl: shareMessage.data.image
                },
                user: user,
                rule: rule.data.content
            }
        });
    });
};
