// api/lib/apis/invitation.js
/**
 * Created by 沐沐 on 2016-01-20.
 * 邀请码
 */
var request = require('./request');

/**
 * 我邀请的用户列表
 * @param tokenOrUnionid
 * @param page
 * @param pageSize
 * @returns {*}
 */
module.exports.invitationList = function (tokenOrUnionid, page, pageSize) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            page: page,
            pageSize: pageSize,
            method: 'bubugao.mobile.app.reg.invitationcode.list'
        }, tokens);
    });
};

/**
 * 自己的邀请码信息
 * @param tokenOrUnionid
 * @returns {*}
 */
module.exports.invitationGet = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.app.reg.invitationcode.get'
        }, tokens);
    });
};

/**
 * 用户提交邀请码
 * @param tokenOrUnionid
 * @param code
 * @param weichatPic
 * @param weichatName
 * @returns {*}
 */
module.exports.invitationSubmit = function (tokenOrUnionid, code, weichatPic, weichatName) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.app.reg.invitationcode.invite',
            code: code,
            weichatPic: weichatPic,
            weichatName: weichatName
        }, tokens);
    });
};

/**
 * 用户扫描二维码
 * @param openid
 * @param code
 * @param weichatPic
 * @param weichatName
 * @param mobile
 * @returns {promise}
 */
module.exports.invitationScan = function (openid, code, weichatPic, weichatName, mobile) {
    return request.get({
        method: 'bubugao.mobile.app.reg.invitationcode.scan',
        openid: openid,
        code: code,
        mobile: mobile,
        weichatPic: weichatPic,
        weichatName: weichatName
    });
};

/**
 * 分享时提交
 * @param tokenOrUnionid
 * @param weichatPic
 * @param weichatName
 * @returns {*}
 */
module.exports.shareSubmit = function (tokenOrUnionid, weichatPic, weichatName) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.app.reg.invitationcode.share',
            weichatPic: weichatPic,
            weichatName: weichatName
        }, tokens)
    })
};

/**
 * 分享列表
 * @param code
 * @param page
 * @param pageSize
 * @returns {promise}
 */
module.exports.shareList = function (code, page, pageSize) {
    return request.get({
        method: 'bubugao.mobile.app.reg.invitationcode.get.bycode',
        page: page,
        pageSize: pageSize,
        code: code
    });
};
