// api/lib/apis/kill.js
/**
 * Created by 沐沐 on 2015-09-14.
 */

var q = require('q');
var request = require('./request');
var error = require('../error');

/**
 * 秒杀商品列表
 * @param activityType
 * @param beginTime
 * @param page
 * @param pageSize
 * @param areaId
 * @returns {*}
 */
module.exports.gift = function (activityType, beginTime, page, pageSize, areaId) {
    if (beginTime) {
        return request.get({
            method: 'bubugao.mobile.gift.search',
            isfilterNoBegin: 1,
            isfilterNoStore: 1,
            activityType: activityType || 0,
            areaId: areaId || 0,
            page: page,
            pageSize: pageSize,
            beginTime: beginTime
        });
    } else {
        return q.fcall(function () {
            throw new error.ApiError('9999', '请输入商品抢购时间！');
        });
    }
};

/**
 * 秒杀时间点
 * @param activityType
 * @param areaId
 * @returns {*}
 */
module.exports.points = function (activityType, areaId) {
    return request.get({
        method: 'bubugao.mobile.gift.getTimes',
        activityType: activityType,
        areaId: areaId || 0
    }).then(function (times) {
        return times.data.map(function (e, i) {
            return e.begintime;
        });
    });
};

/**
 * 登记提醒手机号码
 * @param mobile
 * @returns {promise}
 */
module.exports.mobile = function (mobile) {
    return request.get({
        method: 'bubugao.mobile.gift.recodeMobile',
        mobile: mobile
    });
};


/**
 * 男神女神按城市查询
 * @param city 长沙：cs，湘潭：xt，岳阳：yy
 * @returns {promise}
 */
module.exports.city = function (city) {
    var area = {
        cs: '430100000000',
        xt: '430300000000',
        yy: '430600000000'
    };
    return request.get({
        method: 'bubugao.mobile.gift.activityCity',
        areaId: area[city] || area['cs'],
        activityType: 1
    });
};
