// api/lib/apis/login.js
/**
 * Created by 沐沐 on 2015-07-29.
 * 登陆部分
 */

var weixin = require('../weixin/index');
var request = require('./request');
var error = require('../error');

/**
 * 绑定页面数据
 * @param code
 * @returns {*}
 */
module.exports.index = function (code) {
    return weixin.open_id(code);
};

/**
 * 注册协议
 * @returns {promise}
 */
module.exports.licenses = function (code) {
    var _code=code||'zcxy';
    return request.get({
        method: 'bubugao.mobile.licenses.get',
        code: _code
    });
};

/**
 * App2.6登录接口
 * @param unionid 微信unionid
 * @param openid 微信openid
 * @param mobile 手机号码
 * @param captcha 短信验证码
 * @param channel 平台代码 （梅溪湖项目新增参数）
 * @returns {*}
 */
module.exports.login = function (unionid, openid, mobile, captcha, channel) {
    if (unionid && openid && unionid !== 'undefined' && openid !== 'undefined') {
        return request.get({
            method: 'bubugao.mobile.register.login',
            mobile: mobile,
            captcha: captcha,
            openid: openid,
            channel:channel,
            appId: 1,
            bundleId: 'weixin',
            identify: unionid
        }).then(function (data) {
            return request.get({
                method: 'bubugao.mobile.user.weixin.binding',
                mobile: mobile,
                accountId: data.data.accountId,
                unionId: unionid,
                openId: openid
            }).then(function (bindResult) {
                bindResult.data = bindResult.data || {};
                bindResult.data.newUser = data.data.newUser;
                return bindResult;
            });
        });
    } else {
        throw(new error.ApiError('9999', '无法获取微信身份信息,请联系客服'));
    }
};
/**
 * 20161027 o2o更新后新登录接口
 * @param unionid 微信unionid
 * @param openid 微信openid
 * @param mobile 手机号码
 * @param captcha 短信验证码
 * @param channel 平台代码 （梅溪湖项目新增参数）
 * @param loginKey
 * @param tokenSession 重新登录所需字段tokenid
 * @returns {*}
 */
module.exports.loginV2 = function (unionid, openid, mobile, captcha, channel,loginKey,tokenSession) {
    if (unionid && openid && unionid !== 'undefined' && openid !== 'undefined') {
        return request.get({
            method: 'bubugao.mobile.register.login.v2',
            mobile: mobile,
            captcha: captcha,
            openid: openid,
            channel:channel,
            appId: 1,
            bundleId: 'weixin',
            identify: unionid,
            loginKey:loginKey,
            tokenSession:tokenSession
        }).then(function (data) {
            return request.get({
                method: 'bubugao.mobile.user.weixin.binding',
                mobile: mobile,
                accountId: data.data.accountId,
                unionId: unionid,
                openId: openid
            }).then(function (bindResult) {
                bindResult.data = bindResult.data || {};
                bindResult.data.newUser = data.data.newUser;
                bindResult.data.babyNewUser = data.data.babyNewUser;
                bindResult.data.channel = data.data.channel;
                bindResult.data.accountId = data.data.accountId;
                return bindResult;
            });
        });
    } else {
        throw(new error.ApiError('9999', '无法获取微信身份信息,请联系客服'));
    }
};

/**
 * 更新渠道id
 * @param unionid
 * @param channel
 * @returns {*}
 */
module.exports.updateChannel = function (unionid,channel) {
    return request.get({
        method: 'bubugao.member.channel.update',
        unionid:unionid,
        channel:channel
    });
};
