// api/lib/apis/message.js
/**
 * Created by 沐沐 on 2015-10-14.
 * 消息接口
 */

var request = require('./request');

/**
 * 消息分类总览
 * @param tokenOrUnionid
 * @returns {promise}
 */
module.exports.omitMessages = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.message.omitContent.get'
        }, tokens);
    });
};

/**
 * 分类消息列表查询
 * @param tokenOrUnionid
 * @param page
 * @param pageSize
 * @param type
 * @returns {promise}
 */
module.exports.messageList = function (tokenOrUnionid, page, pageSize, type) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.message.list.get',
            page: page,
            pageSize: pageSize,
            type: type
        }, tokens);
    });
};

/**
 * 消息数量统计
 * @param tokenOrUnionid
 * @returns {promise}
 */
module.exports.messageCount = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.message.count.get'
        }, tokens);
    });
};

/**
 * 消息已读标记
 * @param tokenOrUnionid
 * @param id
 * @param type
 * @returns {promise}
 */
module.exports.mark = function (tokenOrUnionid, type, id) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.message.mark',
            type: type,
            id: id
        }, tokens);
    });
};

/**
 * 清除消息
 * @param tokenOrUnionid
 * @param id
 * @param type
 * @returns {promise}
 */
module.exports.clean = function (tokenOrUnionid, type, id) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.message.clean',
            type: type,
            id: id
        }, tokens);
    });
};

/**
 * 新增消息（梅溪湖项目新增）
 * @param tokenOrUnionid
 * @param {int} msgType = [1, "交易 " | 2, "优惠" | 3, "活动" | 4, "系统广播]
 * @param {string} msgTitle
 * @param {string} msg
 * @returns {*}
 */
module.exports.add = function (tokenOrUnionid, msgType, msgTitle, msg) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.message.add',
            msgType: msgType,
            msgTitle: msgTitle,
            msg: msg
        }, tokens);
    });
};
