// api/lib/apis/order.js
/**
 * Created by 沐沐 on 2015-10-14.
 * 订单接口
 */
var request = require('./request');

/**
 * 礼品订单数目
 * @param tokenOrUnionid
 * @returns {*}
 */
module.exports.giftOrderCount = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.gift.order.count'
        }, tokens);
    });
};

/**
 * 商城订单数量
 * @param tokenOrUnionid
 * @returns {*}
 */
module.exports.mallOrderCount = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.mall.order.count.global'
        }, tokens);
    })
};

/**
 * 小店订单数量
 * @param tokenOrUnionid
 * @returns {*}
 */
module.exports.commodityOrderCount = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.local.commodity.userorder.count'
        }, tokens);
    });
};

/**
 * 商城订单列表
 * @param tokenOrUnionid
 * @param {int} pageNo 页码
 * @param {int} pageSize 一页多少条
 * @param {int} payStatus=[0|1] 0：待付款；1：待发货
 * @returns {*}
 */
module.exports.mallOrderSearch = function (tokenOrUnionid, pageNo, pageSize, payStatus) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.mall.order.search.global',
            pageNo: pageNo,
            pageSize: pageSize,
            payStatus: payStatus
        }, tokens);
    });
};

/**
 * 礼品订单列表
 * @param tokenOrUnionid
 * @param {int} page 页码
 * @param {int} pageSize 一页多少条
 * @param {int} status = [0|1|2] 0：全部 ； 1：待发货； 2：待收货
 * @returns {*}
 */
module.exports.giftOrderSearch = function (tokenOrUnionid, page, pageSize, status) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.gift.orderlist',
            page: page,
            pageSize: pageSize,
            status: status
        }, tokens)
    });
};

/**
 * 生活订单列表
 * @param tokenOrUnionid
 * @param {int} page 页码
 * @param {int} pageSize 一页多少条
 * @param {int} orderStatus = [0|2|3|6] 0：全部；2：未消费；3：退款中；6：已过期
 * @returns {*}
 */
module.exports.commodityOrderSearch = function (tokenOrUnionid, page, pageSize, orderStatus) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.local.commodity.order.list',
            page: page,
            pageSize: pageSize,
            orderStatus: orderStatus
        }, tokens);
    });
};

/**
 * 商城订单详情
 * @param tokenOrUnionid
 * @param {string} orderId 订单编号
 * @returns {*}
 */
module.exports.mallOrderInfo = function (tokenOrUnionid, orderId) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.mall.order.info.get',
            orderId: orderId
        }, tokens);
    });
};

/**
 * 礼品订单详情
 * @param tokenOrUnionid
 * @param {string} orderId 订单编号
 * @returns {*}
 */
module.exports.giftOrderInfo = function (tokenOrUnionid, orderId) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.gift.orderdetail',
            id: orderId
        }, tokens);
    });
};

/**
 * 生活订单详情
 * @param tokenOrUnionid
 * @param orderId
 * @returns {*}
 */
module.exports.commodityOrederInfo = function (tokenOrUnionid, orderId) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.local.commodity.userorder.details',
            orderId: orderId
        }, tokens);
    });
};

/**
 *  申请订单退款（3.0新增接口）
 * @param tokenOrUnionid
 * @param type
 * @param orderId
 * @param reason
 * @param remark
 * @param image1
 * @param image2
 * @param image3
 * @returns {{token: string, secret: string}}
 */
module.exports.orderCancel = function (tokenOrUnionid, type, orderId, reason, remark, image1, image2, image3) {
    return request.getTokens(tokenOrUnionid, function (tokens) {
        return request.postAuth({
            method: 'bubugao.mobile.gift.order.cancel.apply',
            type: type,
            orderId: orderId,
            reason: reason,
            remark: remark
        }, {
            image1: image1,
            image2: image2,
            image3: image3
        }, tokens, true);
    });
};

/**
 * 订单退款详情（3.0新增接口）
 * @param orderId
 * @returns {promise}
 */
module.exports.orderCancelDetail = function(orderId){
    return request.get({
        method:'bubugao.mobile.gift.order.cancel.apply.detail',
        orderId:orderId
    });
};
