// api/lib/apis/paypass.js
/**
 * Created by 沐沐 on 2015-12-18.
 */
var request = require('./request');

module.exports.checkPayPass = function (unionidOrTokens, payPwd) {
    return request.getTokens(unionidOrTokens).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.bbgpay.member.check.password.wechat',
            payPwd: payPwd
        },tokens);
    });
};

/**
 * 生成付款码
 * @param unionidOrTokens
 * @param payamt 金额
 * @param payPwd
 * @returns {*}
 */
module.exports.barcode = function (unionidOrTokens, payamt, payPwd) {
    return request.getTokens(unionidOrTokens).then(function (tokens) {
        return module.exports.checkPayPass(tokens, payPwd).then(function () {
            return request.getAuth({
                method: 'bubugao.mobile.bbgpay.member.gen.barcode',
                payamt: payamt
            }, tokens);
        });
    });
};

/**
 * 付款码状态
 * @param unionidOrTokens
 * @param barcode
 * @returns {*}
 */
module.exports.barcodeStatus = function (unionidOrTokens, barcode) {
    return request.getTokens(unionidOrTokens).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.bbgpay.member.get.barcode.status',
            barcode: barcode
        }, tokens);
    });
};

/**
 * 查询订单
 * @param orderId
 * @returns {promise}
 */
module.exports.apporder = function (orderId) {
    return request.get({
        method: 'bubugao.mobile.apporder.get',
        orderId: orderId
    });
};

/**
 * 用红包支付接口
 * @param unionidOrTokens
 * @param storeId
 * @param orderNo
 * @param txamt
 * @param bizType
 * @param payPwd
 * @param cipherKey
 * @param payType
 * @returns {*}
 */
module.exports.apporderSave = function (unionidOrTokens, storeId, orderNo, txamt, bizType, payPwd, cipherKey, payType) {
    return request.getTokens(unionidOrTokens).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.member.deal.save',
            storeId: storeId,
            orderNo: orderNo,
            txamt: txamt,
            //bizType: bizType,
            payPwd: payPwd,
            cipherKey: cipherKey,
            payType: payType
        }, tokens);
    });
};
