// api/lib/apis/store.js
/**
 * Created by 沐沐 on 2015-11-05.
 */

var request = require('./request');


/**
 * 按地区查询门店
 * @param areaid 地区id
 * @param gaobi 是否支持高币 可为空
 * @param categoryId 商家分类ID 可为空
 * @param giftId 礼品ID 可为空
 * @returns {promise}
 */
module.exports.storeArea = function (areaid, gaobi, categoryId, giftId) {
    return request.get({
        method: 'bubugao.mobile.store.area',
        areaId: areaid,
        gaobi: gaobi,
        categoryId: categoryId,
        giftId: giftId
    })
};

/**
 * 查询指定范围内的门店
 * @param distance 查询范围（米）
 * @param longitude 中心经度
 * @param latitude 中心纬度
 * @param gaobi 是否支持高币 可为空
 * @param categoryId 商家分类ID 可为空
 * @returns {promise}
 */
module.exports.storeDistance = function (distance, longitude, latitude, gaobi, categoryId) {
    return request.get({
        method: 'bubugao.mobile.store.distance',
        distance: distance,
        longitude: longitude,
        latitude: latitude,
        gaobi: gaobi,
        categoryId: categoryId
    });
};

/**
 * 门店详情
 * @param storeId
 * @returns {promise}
 */
module.exports.storeDetail = function (storeId) {
    return request.get({
        method: 'bubugao.mobile.store.detail',
        storeId: storeId
    })
};

/**
 * 门店分类
 * @returns {promise}
 */
module.exports.storeCategory = function () {
    return request.get({
        method: 'bubugao.mobile.seller.category.list.get'
    })
};

/**
 * 获取商家下门店客服列表
 * @param areaId 地区id
 * @param online 是否开店
 * @param categoryId 门店类型
 * @returns {promise}
 */
module.exports.storeHotLine = function (areaId, online, categoryId) {
    return request.get({
        method: 'bubugao.mobile.customer.shop.searchonline',
        areaId: areaId,
        online: online,
        categoryId: categoryId
    })
};

/**
 * 获取广告
 * @param zoneId
 * @returns {promise}
 */
module.exports.advert = function (zoneId) {
    return request.get({
        method: 'bubugao.mobile.advert.get',
        zoneId: zoneId
    });
};

/**
 * 门店精彩活动
 * @param page
 * @param pageSize
 * @param storeid
 * @returns {promise}
 */
module.exports.activity=function (page,pageSize,storeid) {
    return request.get({
        method: 'bubugao.mobile.store.activity.search',
        page:page,
        pageSize:pageSize,
        storeid:storeid
    })
};

/**
 * 渠道id换取门店storeid
 * @param channel
 * @returns {*}
 */
module.exports.exchangeStoreId=function (channel) {
    return request.get({
        method: 'bubugao.mobile.storeid.by.code',
        code:channel
    })
};
