// api/lib/apis/thirdParty.js
/**
 * Created by ricopter@qq.com on 2017/1/18.
 *
 */
'use strict';
var q = require('q');
var conf = require('../conf');
var request = require('./request');

/**
 * 伊利授权跳转地址
 * @param unionid
 * @returns {Promise|*}
 */
module.exports.authYiLiUrl = function (unionid) {
  return request.getTokens(unionid).then(function (tokens) {
    return request.get({
      unionId:unionid,
      method: 'bubugao.mobile.meichis.user.redirect'
    }, tokens);
  });
};

/**===============春雨医生 start=================**/

/**
 * 图片上传
 * @param unionid
 * @param type  image、audio
 * @data objects  { image: { file: '/home/tomas/linux.png', content_type: 'image/png' } }//form 表单字段名
 * @returns {Promise}
 */
module.exports.springDoctorFileUpLoad = function (unionid,type,data) {
  return request.getTokens(unionid).then(function (tokens) {
    return request.postAuth({
      type:type,
      method: 'bubugao.mobile.meichis.user.file.upLoad'
    },data,tokens,true);
  });
};

/**
 * 创建问题
 * @param unionid
 * @param type  image、audio
 * @param contentText  提问内容限制十个字以上
 * @param sex          男/女
 * @param birthday  2010-01-01
 * @param imageUrl
 * @param aduioUrl
 * @returns {Promise}
 */
module.exports.springDoctorCreateQuestion = function (unionid,type,contentText,sex,birthday,imageUrl,aduioUrl) {
  return request.getTokens(unionid).then(function (tokens) {
    return request.getAuth({
      type:type,
      contentText:contentText,
      sex:sex,
      birthday:birthday,
      aduioUrl:aduioUrl||undefined,
      imageUrl:imageUrl||undefined,
      method: 'bubugao.mobile.meichis.user.create.question'
    },tokens);
  });
};

/**
 * 问题追问
 * @param unionid
 * @param questionID  问题id
 * @param contentText  提问内容(限制十个字以上)
 * @param imageUrl   图片url(通过上传接口获取)
 * @returns {Promise}
 */
module.exports.springDoctorAskingQuestion = function (unionid,questionID,contentText,imageUrl) {
  return request.getTokens(unionid).then(function (tokens) {
    return request.getAuth({
      questionID:questionID,
      contentText:contentText,
      imageUrl:imageUrl||undefined,
      method: 'bubugao.mobile.meichis.user.question.asking'
    },tokens);
  });
};

/**
 * 查看问题回复
 * @param unionid
 * @param questionID
 * @returns {Promise}
 */
module.exports.springDoctorReplyQuestion = function (unionid,questionID) {
  return request.getTokens(unionid).then(function (tokens) {
    return request.getAuth({
      questionID:questionID,
      method: 'bubugao.mobile.meichis.user.question.reply.query'
    },tokens);
  });
};

/**
 * 关闭问题
 * @param unionid
 * @param questionID
 * @param closeReason
 * @returns {Promise}
 */
module.exports.springDoctorCloseQuestion = function (unionid,questionID,closeReason) {
  return request.getTokens(unionid).then(function (tokens) {
    return request.getAuth({
      questionID:questionID,
      closeReason:closeReason,
      method: 'bubugao.mobile.meichis.user.question.close'
    },tokens);
  });
};
/**
 * 问题评价
 * @param unionid
 * @param questionID  问题id
 * @param evaluation  评价星级 1-5
 * @param content  评语
 * @returns {Promise}
 */
module.exports.springDoctorEvaluateQuestion = function (unionid,questionID,evaluation,content) {
  return request.getTokens(unionid).then(function (tokens) {
    return request.getAuth({
      questionID:questionID,
      evaluation:evaluation,
      content:content,
      method: 'bubugao.mobile.meichis.user.question.evaluate'
    },tokens);
  });
};

/**
 * 查询聊天记录
 * @param unionid
 * @param questionID
 * @returns {Promise}
 */
module.exports.springDoctorQueryQuestionRecord = function (unionid,questionID) {
  return request.getTokens(unionid).then(function (tokens) {
    return request.getAuth({
      questionID:questionID,
      method: 'bubugao.mobile.meichis.user.question.query'
    },tokens);
  });
};

/**
 *  查询问题列表
 * @param unionid
 * @param page     init
 * @param pageSize init
 * @param status 已回复|已关闭|待回复
 * @returns {Promise}
 */
module.exports.springDoctorQueryQuestionList = function (unionid,page,pageSize,status) {
  return request.getTokens(unionid).then(function (tokens) {
    return request.getAuth({
      page:page,
      pageSize:pageSize,
      status:status,
      method: 'bubugao.mobile.meichis.user.query.list'
    },tokens);
  });
};
/**===============春雨医生 end=================**/
