// api/lib/apis/unicom.js
/**
 * Created by 沐沐 on 2015-11-09.
 */

var request = require('./request');

/**
 * 检验联通账号
 * @param tokenOrUnionid
 * @param phone
 * @returns {*}
 */
module.exports.validate = function (tokenOrUnionid, phone) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.app.unicom.integral.validate',
            phone: phone
        }, tokens);
    });
};

/**
 * 查询联通积分
 * @param tokenOrUnionid
 * @param phone
 * @returns {*}
 */
module.exports.integration = function (tokenOrUnionid, phone) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.app.unicom.integral.get',
            phone: phone
        }, tokens);
    });
};

/**
 * 联通积分换云猴红包
 * @param tokenOrUnionid
 * @param phone
 * @param integral
 * @returns {*}
 */
module.exports.exchangeHongbao = function (tokenOrUnionid, phone, integral) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.app.unicom.integral.exchange',
            phone: phone,
            integral: integral
        }, tokens);
    });
};

/**
 * 联通积分换云猴积分
 * @param tokenOrUnionid
 * @param phone
 * @param integral
 * @returns {*}
 */
module.exports.exchangeIntegral = function (tokenOrUnionid, phone, integral) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.app.unicom.integral.exchange.v2',
            phone: phone,
            integral: integral
        }, tokens);
    });
};
