// api/lib/apis/unionpay.js
/**
 * Created by 沐沐 on 2015-11-11.
 */


var request = require('./request');

/**
 * 验证银行卡等级接口(已登录状态)
 * @param tokenOrUnionid
 * @param bankNo
 * @returns {*}
 */
module.exports.verify = function (tokenOrUnionid, bankNo) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.unionpay.member.verify.auth',
            bankNo: bankNo
        }, tokens);
    });
};

/**
 * 查询用户已绑定的有效银联卡接口
 * @param tokenOrUnionid
 * @returns {*}
 */
module.exports.cardList = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        //console.log(tokens);
        return request.getAuth({
            method: 'bubugao.mobile.unionpay.member.cardlist'
        }, tokens).then(function (data) {
            //data.data.cardlist = [
            //    {
            //        unioncardNo:'622202190100961'
            //    }
            //];
            if (data.data && data.data.cardlist && data.data.cardlist.length) {
                data.data.cardlist = data.data.cardlist.map(function (e) {
                    var cardNo = e.unioncardNo;
                    var mixedCardNo = '';
                    for (var i = 0; i < cardNo.length; i++) {
                        if (i < 4 || i > cardNo.length - 4) {
                            mixedCardNo += cardNo[i];
                        } else if (i >= 4 && i <= cardNo.length - 4) {
                            mixedCardNo += '*';
                        }
                    }
                    e.mixedUnioncardNo = mixedCardNo;
                    return e;
                })
            }
            return data;
        });
    });
};


/**
 * 银联积分云猴红包
 * @param tokenOrUnionid
 * @param integral
 * @param bankAccountNo
 * @returns {*}
 */
module.exports.exchangeHongbao = function (tokenOrUnionid, integral, bankAccountNo) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.unionpay.member.integral.exchange',
            integral: integral,
            bankAccountNo: bankAccountNo
        }, tokens);
    });
};

/**
 * 银联积分换云猴积分
 * @param tokenOrUnionid
 * @param integral
 * @param bankAccountNo
 * @returns {*}
 */
module.exports.exchangeIntegral = function (tokenOrUnionid, integral, bankAccountNo) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.unionpay.member.integral.exchange.v2',
            integral: integral,
            bankAccountNo: bankAccountNo
        }, tokens);
    });
};
