// api/lib/apis/user.js
/**
 * Created by 沐沐 on 2015-07-17.
 * App Api 整合模块
 */
var q = require('q');
var request = require('./request');
var coupons = require('./coupons');
var message = require('./message');
var invitation = require('./invitation');
var weixin = require('../weixin');
var PaySign = require("./paySign");
/**
 * 会员信息
 * @param {{token: string, secret: string}} tokenOrUnionid
 * @returns {promise}
 * @callback {{data: {name: string, sex: string, mobile: string, birthday: string, certificateNo: string,
 * payPass: boolean, qrcode: string, isSiebel: boolean},code: string}}
 */
module.exports.meberInfo = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.local.siebel.member.info'
        }, tokens);
    });
};

/**
 * 获得绑定时填写的手机号码
 * @param {{token: string, secret: string}|string} tokenOrUnionid
 * @returns {promise}
 * @callback {{data: {mobile: string},code: string}}
 */
module.exports.mobile = function (tokenOrUnionid, tokenid) {
    return request.getTokens(tokenOrUnionid,tokenid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.user.mobile.get'
        }, tokens);
    });
};

/**
 * APP2.6 完善个人信息接口
 * @param {string} tokenOrUnionid
 * @param {string} mobile
 * @param {string} idNo
 * @param {string} memberName
 * @returns {promise}
 */
module.exports.sync = function (tokenOrUnionid, mobile, idNo, memberName) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.local.siebel.member.sync',
            mobile: mobile,
            idNo: idNo,
            memberName: memberName
        }, tokens);
    })
};

/**
 * 查询用户会员卡
 * @param tokenOrUnionid
 * @returns {promise}
 */
module.exports.cards = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.local.siebel.card.list'
        }, tokens);
    });
};

/**
 * APP2.6 积分查询
 * @param {{token: string, secret: string}|string} tokenOrUnionid
 * @returns {promise}
 */
module.exports.integration = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.local.siebel.point.get'
        }, tokens);
    });
};


/**
 * APP2.7积分增加接口
 * @param {{token: string, secret: string} | string} tokenOrUnionid
 * @param {string} transType = 'Point Accrucal'
 * @param {string} transFrom = [MTL|YAO|XZJ]
 * @param {int} points
 * @returns {promise}
 */
module.exports.integrationIncr = function (tokenOrUnionid, transType, transFrom, points) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.local.siebel.integral.incr',
            transType: transType,
            transFrom: transFrom,
            points: points
        }, tokens);
    });
};

/**
 * APP2.6 查询用户积分详情
 * @param {string} tokenOrUnionid
 * @param {int} page 请求页码
 * @param {int} pageSize 每页显示条数
 * @return {promise}
 */
module.exports.integrationList = function (tokenOrUnionid, page, pageSize) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.local.siebel.deals.get',
            page: page,
            pageSize: pageSize
        }, tokens);
    })
};

/**
 * 积分转红包
 * @param tokenOrUnionid
 * @param integration
 * @returns {*}
 */
module.exports.transferIntegration = function (tokenOrUnionid, integration) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.integral.transfer.gaobi',
            integral: integration
        });
    });
};


/**
 * 红包查询
 * @param {{token: string, secret: string}|string} tokenOrUnionid
 * @returns {promise}
 */
module.exports.reward = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.bbgpay.member.balance.get'
        }, tokens);
    });
};

/**
 * 查询用户收支明细
 * @param {string} tokenOrUnionid
 * @param {int} type = [1|2] 1：云猴宝；2：云猴红包
 * @param {int} page 请求页码
 * @param {int} pageSize 每页记录条数
 * @param {int} pageFlag = [0|1] 是否分页 0 1
 * @return {promise}
 */
module.exports.rewardList = function (tokenOrUnionid, type, page, pageSize, pageFlag) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.bbgpay.member.gbal.list',
            pageFlag: pageFlag,
            gbType: type,
            pageIndex: page,
            pageSize: pageSize
        }, tokens);
    });
};

/**
 * 按生效未生效查询红包
 * @param tokenOrUnionid
 * @param type 查询类别是未生效，已生效，已失效其值分别是U1,U2,U3
 * @param page 当前页码,从1开始
 * @param pageSize 页大小
 * @param pageFlag 分页，1：不分页
 * @returns {*}
 */
module.exports.rewardListByType = function (tokenOrUnionid, type, page, pageSize, pageFlag) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.bbgpay.member.redpkg.list',
            pageFlag: pageFlag,
            queryType: type,
            pageIndex: page,
            pageSize: pageSize
        }, tokens);
    });
};

/**
 * 码兑优惠券
 * @param tokenOrUnionid
 * @param code
 * @returns {*}
 */
module.exports.hongbaoExchange = function (tokenOrUnionid, code) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.cdkey.code.exchange',
            code: code
        }, tokens);
    });
};

/**
 * 积分红包查询合并接口
 * @param tokenOrUnionid
 * @returns {*}
 */
module.exports.user_account = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return  module.exports.integration(tokens).then(function (integration) {
            return {
                code: '0000',
                data: {
                    isAccountActive: true,
                    integration: integration.data.point,
                    reward: null
                }
            }
        }).fail(function (err) {
            if (err.code == '1010') {
                return module.exports.mobile(tokens).then(function (mobile) {
                    return {
                        code: err.code,
                        data: {
                            isAccountActive: false,
                            integration: null,
                            reward: null,
                            mobile: mobile.data.mobile
                        }
                    };
                });
            } else {
                throw err;
            }
        });
    })
};

module.exports.myCard = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return q.all([
            module.exports.cards(tokens),
            module.exports.meberInfo(tokens),
            invitation.invitationGet(tokens)
        ]).spread(function (cards, member, invitation) {
            return {
                cards: cards.data.length > 0 ? cards.data[0] : null,
                member: member.data,
                invitation: invitation.data
            }
        }).fail(function (err) {
            if (err.code == '1010') {
                return {
                    cards: null,
                    member: null,
                    invitation: null
                };
            } else {
                throw err;
            }
        });
    });
};

/**
 * 查询母婴用户会员卡
 * @param tokenOrUnionid
 * @returns {promise}
 */
module.exports.babyCards = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.local.siebel.card'
        }, tokens);
    });
};
module.exports.myBabyCards = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return q.all([
            module.exports.babyCards(tokens),
            module.exports.meberInfo(tokens),
            invitation.invitationGet(tokens)
        ]).spread(function (cards, member, invitation) {
            return {
                cards: cards.data.length > 0 ? cards.data[0] : null,
                member: member.data,
                invitation: invitation.data
            }
        }).fail(function (err) {
            if (err.code == '1010') {
                return {
                    cards: null,
                    member: null,
                    invitation: null
                };
            } else {
                throw err;
            }
        });
    });
};

/**
 * 完善母婴会员信息
 * @param tokenOrUnionid
 * @param sex
 * @param birthday 2016-12-01
 * @returns {*}
 */
module.exports.updateBabyMember = function (tokenOrUnionid,sex,birthday) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.member.baby.update',
            sex:sex,
            birthday:birthday
        }, tokens);
    });
};
/**
 *
 * @param tokenOrUnionid
 * @returns {*}
 */
module.exports.totalAccount = function (tokenOrUnionid) {
    return request.getTokens(tokenOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.siebel.voucher.push.index'
        }, tokens);
    });
};

/**
 * 会员信息合并接口
 * @param {string} tokenOrUnionid
 * @param {string} openid
 * @returns {promise}
 */
module.exports.center = function (tokenOrUnionid, openid, tokenid) {
    return request.getTokens(tokenOrUnionid, tokenid).then(function (tokens) {
        //console.info(tokens);
        return q.all([
            module.exports.integration(tokens),
            //module.exports.reward(tokens),
            //coupons.couponMall(tokens, '1', 1, 100),
            // coupons.couponMember(tokens, '1', 1, 100),
            coupons.couponSiebel(tokens, '可用', 1, 100),
            module.exports.meberInfo(tokens),
            module.exports.cards(tokens),
            PaySign.getUrlSignParam(tokens)
        ]).spread(function (integration, couponSiebel, member, cards,paySignParam) {
            return {
                isAccountActive: true,
                integration: integration.data.point,
                reward: null,
                //coupon: couponMall.count + couponSiebel.count,
                coupon: couponSiebel.count,
                cards: cards.data.length > 0 ? cards.data[0] : null,
                member: member.data,
                paySignParam:paySignParam
            };
        }).fail(function (err) {
            if (err.code == '1010') {
                return {
                    isAccountActive: false,
                    integration: null,
                    reward: null,
                    coupon: null,
                    cards: null,
                    member: null
                };
            } else {
                //console.error(err);
                throw err;
            }
        }).then(function (info) {

            //  超级app登录绕过微信授权
            if(tokenid && tokenid !== undefined && /superapp/ig.test(tokenid)){
                return q.all([
                    message.messageCount(tokens),
                    module.exports.mobile(tokens),
                ]).spread(function (messageCount, mobile) {
                    info.mobile = mobile.data.mobile;
                    info.message = messageCount.data.count;
                    return info;
                });
            }

            return weixin.access_token().then(function (accessToken) {
                return q.all([
                    message.messageCount(tokens),
                    module.exports.mobile(tokens),
                    weixin.user_info(accessToken.access_token, openid)
                ]).spread(function (messageCount, mobile, weixinUser) {
                    info.mobile = mobile.data.mobile;
                    info.message = messageCount.data.count;
                    info.weixin = weixinUser;
                    return info;
                });
            });
        });
    });
};
