// api/lib/weixin/index.js
/**
 * Created by 沐沐 on 2015-07-20.
 * 微信模块
 */

var https = require('https');
var http = require('http');
var queryString = require('querystring');
var q = require('q');
var conf = require('../conf');
var error = require('../error');
var sign = require('./sign');
var logs = require('logs').logs;
var logsErr = logs.recordCate("err");//记录错误
var needle = require('needle');

var send = function (path) {
    var deferred = q.defer();
    var timeout = conf().api.timeout || 30000;
    var url = 'https://api.weixin.qq.com' + path;
    needle.get(url, {
        timeout: timeout
    }, function (err, res, body) {
        if (err) {
            conf().weixin.printLog && logsErr.error('返回微信报错[' + path + ']:' + err.message);
            var syerr = new error.WeChatError(err.code, err.message, path);
            syerr.data = err;
            deferred.reject(syerr);
        } else {
            if (typeof body === 'string') {
                body = JSON.parse(body);
            }
            if (res.statusCode === 200) {
                if (body.errcode) {
                    conf().weixin.printLog && logsErr.error('返回微信报错[' + body.errcode + ']:' + body.errmsg);
                    deferred.reject(new error.WeChatError(body.errcode, body.errmsg, path));
                } else {
                    conf().weixin.printLog && console.log('返回结果[' + path + ']:' + JSON.stringify(body));
                    deferred.resolve(body);
                }
            } else {
                conf().weixin.printLog && logsErr.error('返回微信服务器报错[' + res.statusCode + ']:' + res.statusMessage);
                deferred.reject(new error.SystemError(res.statusCode, res.statusMessage, path));
            }
        }
    });

    //https.get({
    //    host: 'api.weixin.qq.com',
    //    path: path
    //}, function (res) {
    //    var chunk = '';
    //    res.setEncoding('utf8');
    //    res.on('data', function (data) {
    //        chunk += data;
    //    });
    //    res.on('end', function () {
    //        if (res.statusCode === 200) {
    //            var data = JSON.parse(chunk);
    //            if (data.errcode) {
    //                conf().weixin.printLog && logsErr.error('返回微信报错[' + data.errcode + ']:' + data.errmsg);
    //                deferred.reject(new error.WeChatError(data.errcode, data.errmsg, path));
    //            } else {
    //                conf().weixin.printLog && console.log('返回结果:' + JSON.stringify(data));
    //                deferred.resolve(data);
    //            }
    //        } else {
    //            conf().weixin.printLog && logsErr.error('返回微信服务器报错[' + res.statusCode + ']:' + res.statusMessage);
    //            deferred.reject(new error.SystemError(res.statusCode, res.statusMessage));
    //        }
    //
    //    });
    //}).on('error', function (error) {
    //    conf().weixin.printLog && logsErr.error('返回微信报错:' + error.message);
    //    deferred.reject(error);
    //});
    return deferred.promise;
};

var sendSelf = function (path) {
    var deferred = q.defer();
    var timeout = conf().api.timeout || 30000;
    needle.get('http://' + conf().api.webHost + path, {
        timeout: timeout
    }, function (err, res, body) {
        if (err) {
            deferred.reject(err);
        } else {
            if (res.statusCode === 200) {
                conf().weixin.printLog && console.log('返回结果:' + JSON.stringify(body));
                deferred.resolve(body);
            } else {
                conf().weixin.printLog && logsErr.error('线上返回报错:' + body.message);
                deferred.reject(new error.SystemError(body.code, body.message));
            }
        }
    });
    //http.get('http://' + conf().api.webHost + path, function (res) {
    //    var chunk = '';
    //    res.setEncoding('utf8');
    //    res.on('data', function (data) {
    //        chunk += data;
    //    });
    //    res.on('end', function () {
    //        var data = JSON.parse(chunk);
    //        if (res.statusCode === 200) {
    //            conf().weixin.printLog && console.log('返回结果:' + JSON.stringify(data));
    //            deferred.resolve(data);
    //        } else {
    //            conf().weixin.printLog && logsErr.error('线上返回报错:' + data.message);
    //            deferred.reject(new error.SystemError(data.code, data.message));
    //        }
    //    });
    //}).on('error', function (err) {
    //    deferred.reject(err);
    //});
    return deferred.promise;
};

/**
 * 从微信服务器读取token
 * @returns {*}
 */
var getToken = function () {
    if (conf().weixin.tokenDebug) {
        conf().weixin.printLog && console.log('向线上服务请求access_token');
        //todo 安全问题
        return sendSelf('/api/weixin/token');
    } else {
        conf().weixin.printLog && console.log('向微信服务器请求access_token');
        return send('/cgi-bin/token?grant_type=client_credential&appid=' + conf().weixin.AppID + '&secret=' + conf().weixin.AppSecret).then(saveToken);
    }
};

/**
 * 本地保存微信token
 * @param token
 * @returns {*}
 */
var tokenCache = null;
var saveToken = function (token) {
    token.createTime = parseInt(new Date().getTime() / 1000);
    tokenCache = token;
    return token;
};

/**
 * 先尝试从本地读取微信token失败后从线上读取
 * @returns {*|promise}
 */
var readToken = function () {
    if (conf().weixin.tokenDebug) {
        conf().weixin.printLog && console.log('向线上服务请求access_token');
        return sendSelf('/api/weixin/token').then(saveToken);
    } else {
        var deferred = q.defer();
        if (tokenCache) {
            if (tokenCache.createTime) {
                var now = parseInt(new Date().getTime() / 1000);
                var time = now - tokenCache.createTime;
                if (time > 3600) {
                    getToken().then(deferred.resolve);
                } else {
                    tokenCache.left = time;
                    deferred.resolve(tokenCache);
                }
            } else {
                getToken().then(deferred.resolve);
            }
        } else {
            getToken().then(deferred.resolve);
        }
        return deferred.promise;
    }
};

var getTicket = function () {
    return readToken().then(function (token) {
        if (conf().weixin.tokenDebug) {
            conf().weixin.printLog && console.log('向线上服务请求ticket');
            return sendSelf('/api/weixin/ticket');
        } else {
            conf().weixin.printLog && console.log('向微信服务器请求ticket');
            return send('/cgi-bin/ticket/getticket?access_token=' + token.access_token + '&type=jsapi').then(saveTicket);
        }
    });
};

var ticketCache = null;
var saveTicket = function (ticket) {
    ticket.createTime = parseInt(new Date().getTime() / 1000);
    ticketCache = ticket;
    return ticket;
};

var readTicket = function () {
    if (conf().weixin.tokenDebug) {
        conf().weixin.printLog && console.log('向线上服务请求ticket');
        return sendSelf('/api/weixin/ticket').then(saveTicket);
    } else {
        var deferred = q.defer();
        if (ticketCache) {
            if (ticketCache.createTime) {
                var now = parseInt(new Date().getTime() / 1000);
                var time = now - ticketCache.createTime;
                if (time > 3600) {
                    getTicket().then(deferred.resolve);
                } else {
                    ticketCache.left = time;
                    deferred.resolve(ticketCache);
                }
            } else {
                getTicket().then(deferred.resolve);
            }
        } else {
            getTicket().then(deferred.resolve);
        }
        return deferred.promise;
    }
};

//输出获取微信token方法
module.exports.getToken = readToken;
//输出获取微信票据方法
module.exports.getTicket = readTicket;

/**
 * 获取全局access_token
 */
module.exports.access_token = function () {
    return readToken();
};

/**
 * 微信js_api签名方法
 * @param url
 * @returns {sign}
 */
module.exports.sign = function (url) {
    return readTicket().then(function (ticket) {
        var signStr = sign(ticket.ticket, url);
        return {
            appId: conf().weixin.AppID,
            timestamp: signStr.timestamp,
            nonceStr: signStr.nonceStr,
            signature: signStr.signature,
            url: signStr.url
        };
    });
};


/**
 * 网页授权获取open_id
 * @param code
 */
module.exports.open_id = function (code) {
    if (conf().weixin.ignore) {
        return q.fcall(function () {
            return {
                openid: conf().weixin.testCode.openid,
                unionid: conf().weixin.testCode.unionid
            };
        });
    } else {
        return send('/sns/oauth2/access_token?appid=' + conf().weixin.AppID +
            '&secret=' + conf().weixin.AppSecret + '&code=' + code + '&grant_type=authorization_code');
    }
};

/**
 * 获取用户信息
 * @param access_token
 * @param open_id
 * @param repeat
 */
module.exports.user_info = function (access_token, open_id, repeat) {
    if (conf().weixin.ignore) {
        return q.fcall(function () {
            return {
                openid: conf().weixin.testCode.openid,
                unionid: conf().weixin.testCode.unionid
            };
        });
    } else {
        return send('/cgi-bin/user/info?access_token=' + access_token + '&openid=' + open_id + '&lang=zh_CN').then(null, function (err) {
            if (err instanceof error.WeChatError && repeat) {
                return getToken().then(function (data) {
                    return module.exports.user_info(data.access_token, open_id);
                });
            } else {
                throw err;
            }
        });
    }
};

/**
 * 网页授权获取用户信息
 * @param access_token
 * @param open_id
 */
module.exports.user_info_temp = function (access_token, open_id) {
    return send('/sns/userinfo?access_token=' + access_token + '&openid=' + open_id + '&lang=zh_CN');
};

var url = require('url');
module.exports.refresh = function (req) {
    var path = url.parse(req.originalUrl).pathname;
    return module.exports.getOpenUrl(path, req.query);
};

/**
 * 转向微信验证地址
 * @param path
 * @param params
 * @param not
 * @returns {string}
 */
module.exports.getOpenUrl = function (path, params, not) {
    var use = {};
    params = params || {};
    not = not || [];
    not.push('code', 'state');
    Object.keys(params).map(function (key) {
        if (not.indexOf(key) < 0) {
            use[key] = params[key];
        }
    });
    var query = queryString.stringify(use, null, null,
        {encodeURIComponent: queryString.unescape});
    var url = query ? path + '?' + query : path;
    if (conf().weixin.ignore) {
        return url + ( query ? '&' : '?') + 'code=' + conf().weixin.testCode.openid;
    } else {
        //logsErr.info(print); //用户同意授权，获取code
        return 'https://open.weixin.qq.com/connect/oauth2/authorize?appid=' + conf().weixin.AppID + '&redirect_uri=' + encodeURIComponent('http://' + conf().api.useHost + url) + '&response_type=code&' +
            'scope=snsapi_base&connect_redirect=1&state=1#wechat_redirect';
    }
};
