// api/lib/easy.js
/**
 * Created by 沐沐 on 2016-04-13.
 * 容易网对接部分
 */

var needle = require('needle');
var q = require('q');
var conf = require('./conf');
var request = require('./apis/request');
var error = require('./error');
var logs = require('logs').logs;
var logsErr = logs.recordCate("err");//记录错误

/**
 *
 * @param accountId
 * @returns {*}
 */
var esay_token = function (accountId) {
    var deferred = q.defer();
    var path = '/api/bubugao/token';
    var url = 'http://' + conf().easy.host + (conf().easy.port ? ':' + conf().easy.port : '') +
        path + '?accountid=' + accountId;

    needle.get(url, function (err, res, body) {
        if (err) {
            conf().easy.printLog && logsErr.error('返回网络报错[' + path + ']:' + err.message);
            var syerr = new error.SystemError(err.code, err.message, path);
            syerr.data = err;
            deferred.reject(syerr);
        } else {
            if (res.statusCode === 200) {
                if (!body.meta.status) {
                    conf().easy.printLog && console.log('返回结果[' + path + ']:' + JSON.stringify(body));
                    deferred.resolve({
                        code: '0000',
                        message: body.meta.msg,
                        data: {
                            token: body.result
                        }
                    });
                } else {
                    conf().easy.printLog && logsErr.error('返回API业务报错[' + body.meta.status + '][' + path + ']:' + body.meta.msg);
                    deferred.reject(new error.ApiError(body.meta.status, body.meta.msg, body.result));
                }
            } else {
                conf().easy.printLog && logsErr.error('返回API服务器报错[' + res.statusCode + '][' + path + ']:' + res.statusMessage);
                deferred.reject(new error.SystemError(res.statusCode, res.statusMessage, path));
            }
        }
    });

    return deferred.promise;
};

/**
 *
 * @param tokensOrUnionid
 * @returns {*}
 */
module.exports.easy_token = function (tokensOrUnionid) {
    return request.getTokens(tokensOrUnionid).then(function (tokens) {
        return request.getAuth({
            method: 'bubugao.mobile.user.info'
        }, tokens);
    }).then(function (data) {
        // console.log(data.data.accountId);
        return esay_token(data.data.accountId);
    });
};
